### PR TITLE
Implement ToString on Records (RFC FS-1022)

### PR DESCRIPTION
**Pull request by liboz**
This extends https://github.com/Microsoft/visualfsharp/pull/1589 to Record types and is an implementation of [RFC FS-1022](https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1022-override-ToString-for-discriminated-unions-and-records.md)

Still need to fix tests.

### CODE REVIEW

File: src/fsharp/IlxGen.fs [hunk 1]
```diff
@@ -6062,6 +6062,46 @@ and GenAbstractBinding cenv eenv tref (vref:ValRef) =
...
+                                               // rest follow from 'StringFormat<T>'
+                                               GenUnitTy cenv eenv m  
+                                               cenv.g.ilg.typ_String 
+                                               cenv.g.ilg.typ_String 
+                                               ilThisTy],[])
+               // Instantiate with our own type
+               let sprintfMethSpec = mkILMethSpec(sprintfMethSpec.MethodRef,AsObject,[],[funcTy])
+               // Here's the body of the method. Call printf, then invoke the function it returns
+               let callInstrs = EraseClosures.mkCallFunc cenv.g.ilxPubCloEnv (fun _ -> 0us) eenv.tyenv.Count Normalcall (Apps_app(ilThisTy, Apps_done cenv.g.ilg.typ_String))
+               let ilMethodDef = mkILNonGenericVirtualMethod ("ToString",ILMemberAccess.Public,[],
```
Comment (dsyme) [inline src/fsharp/IlxGen.fs]: Can you share this code with the Union case?

File: src/fsharp/IlxGen.fs [hunk 2]
```diff
@@ -6418,7 +6458,9 @@ and GenTypeDef cenv mgbuf lazyInitInfo eenv m (tycon:Tycon) =
                  // Records that are value types do not create a default constructor with CLIMutable or ComVisible
                  if not isStructRecord && (isCLIMutable || (TryFindFSharpBoolAttribute cenv.g cenv.g.attrib_ComVisibleAttribute tycon.Attribs = Some true)) then
                      yield mkILSimpleStorageCtor(None, Some cenv.g.ilg.typ_Object.TypeSpec, ilThisTy, [], reprAccess) 
-
+                 
+                 if (not <| tycon.HasMember cenv.g "ToString" []) then
```
Comment (dsyme) [inline src/fsharp/IlxGen.fs]: We generally don't use the backpipe operator in the compiler (except ``xyz <| (fun a -> ...)`` now and then)

**Comment by dsyme** [comment]
@liboz Please test struct records, thanks

**Comment by KevinRansom** [comment]
Nice ...

Thank you

**Comment by abelbraaksma** [comment]
New location of RFC: https://github.com/fsharp/fslang-design/blob/main/FSharp-4.1b/FS-1022-override-ToString-for-discriminated-unions-and-records.md

**Pull request merged**